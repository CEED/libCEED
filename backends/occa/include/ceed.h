#ifndef _OCCA_INCLUDE_CEED_H_
#define _OCCA_INCLUDE_CEED_H_
// Phony header to include when compiling OKL
#endif