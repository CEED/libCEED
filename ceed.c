// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
// reserved. See files LICENSE and NOTICE for details.
//
// This file is part of CEED, a collection of benchmarks, miniapps, software
// libraries and APIs for efficient high-order finite element and spectral
// element discretizations for exascale applications. For more information and
// source code availability see http://github.com/ceed.
//
// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
// a collaborative effort of two U.S. Department of Energy organizations (Office
// of Science and the National Nuclear Security Administration) responsible for
// the planning and preparation of a capable exascale ecosystem, including
// software, applications, hardware, advanced system engineering and early
// testbed platforms, in support of the nation's exascale computing imperative.

#define _POSIX_C_SOURCE 200112
#include <ceed-impl.h>

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/// @cond DOXYGEN_SKIP
static CeedRequest ceed_request_immediate;
static CeedRequest ceed_request_ordered;

static struct {
  char prefix[CEED_MAX_RESOURCE_LEN];
  int (*init)(const char *resource, Ceed f);
  unsigned int priority;
} backends[32];
static size_t num_backends;
/// @endcond

/// @file
/// Implementation of core components of Ceed library
///
/// @defgroup Ceed Ceed: core components
/// @{

/// Request immediate completion
///
/// This predefined constant is passed as the \ref CeedRequest argument to
/// interfaces when the caller wishes for the operation to be performed
/// immediately.  The code
///
/// @code
///   CeedOperatorApply(op, ..., CEED_REQUEST_IMMEDIATE);
/// @endcode
///
/// is semantically equivalent to
///
/// @code
///   CeedRequest request;
///   CeedOperatorApply(op, ..., &request);
///   CeedRequestWait(&request);
/// @endcode
///
/// @sa CEED_REQUEST_ORDERED
CeedRequest *const CEED_REQUEST_IMMEDIATE = &ceed_request_immediate;

/**
  Request ordered completion

  This predefined constant is passed as the \ref CeedRequest argument to
  interfaces when the caller wishes for the operation to be completed in the
  order that it is submitted to the device.  It is typically used in a construct
  such as

  @code
    CeedRequest request;
    CeedOperatorApply(op1, ..., CEED_REQUEST_ORDERED);
    CeedOperatorApply(op2, ..., &request);
    // other optional work
    CeedWait(&request);
  @endcode

  which allows the sequence to complete asynchronously but does not start
  `op2` until `op1` has completed.

  @fixme The current implementation is overly strict, offering equivalent
  semantics to CEED_REQUEST_IMMEDIATE.

  @sa CEED_REQUEST_IMMEDIATE
 */
CeedRequest *const CEED_REQUEST_ORDERED = &ceed_request_ordered;

/// Error handling implementation; use \ref CeedError instead.
int CeedErrorImpl(Ceed ceed, const char *filename, int lineno, const char *func,
                  int ecode, const char *format, ...) {
  va_list args;
  va_start(args, format);
  if (ceed) return ceed->Error(ceed, filename, lineno, func, ecode, format, args);
  return CeedErrorAbort(ceed, filename, lineno, func, ecode, format, args);
}

/// Error handler that returns without printing anything.
///
/// Pass this to CeedSetErrorHandler() to obtain this error handling behavior.
///
/// @sa CeedErrorAbort
int CeedErrorReturn(Ceed ceed, const char *filename, int lineno,
                    const char *func, int ecode, const char *format,
                    va_list args) {
  return ecode;
}

/// Error handler that prints to stderr and aborts
///
/// Pass this to CeedSetErrorHandler() to obtain this error handling behavior.
///
/// @sa CeedErrorReturn
int CeedErrorAbort(Ceed ceed, const char *filename, int lineno,
                   const char *func, int ecode,
                   const char *format, va_list args) {
  fprintf(stderr, "%s:%d in %s(): ", filename, lineno, func);
  vfprintf(stderr, format, args);
  fprintf(stderr, "\n");
  abort();
  return ecode;
}

/// Set error handler
///
/// A default error handler is set in CeedInit().  Use this function to change
/// the error handler to CeedErrorReturn(), CeedErrorAbort(), or a user-defined
/// error handler.
int CeedSetErrorHandler(Ceed ceed,
                        int (eh)(Ceed, const char *, int, const char *,
                                 int, const char *, va_list)) {
  ceed->Error = eh;
  return 0;
}

/**
  Register a Ceed backend

  @param prefix Prefix of resources for this backend to respond to.  For
                example, the reference backend responds to "/cpu/self".
  @param init   Initialization function called by CeedInit() when the backend
                is selected to drive the requested resource.
  @param priority Integer priority.  Lower values are preferred in case the
                  resource requested by CeedInit() has non-unique best prefix
                  match.
  @return 0 on success
 */
int CeedRegister(const char *prefix,
                 int (*init)(const char *, Ceed), unsigned int priority) {
  if (num_backends >= sizeof(backends) / sizeof(backends[0])) {
    return CeedError(NULL, 1, "Too many backends");
  }
  strncpy(backends[num_backends].prefix, prefix, CEED_MAX_RESOURCE_LEN);
  backends[num_backends].init = init;
  backends[num_backends].priority = priority;
  num_backends++;
  return 0;
}

/// Allocate an array on the host; use CeedMalloc()
///
/// Memory usage can be tracked by the library.  This ensures sufficient
/// alignment for vectorization and should be used for large allocations.
///
/// @param n Number of units to allocate
/// @param unit Size of each unit
/// @param p Address of pointer to hold the result.
/// @sa CeedFree()
int CeedMallocArray(size_t n, size_t unit, void *p) {
  int ierr = posix_memalign((void **)p, CEED_ALIGN, n*unit);
  if (ierr)
    return CeedError(NULL, ierr,
                     "posix_memalign failed to allocate %zd members of size %zd\n", n, unit);
  return 0;
}

/// Allocate a cleared (zeroed) array on the host; use CeedCalloc()
///
/// Memory usage can be tracked by the library.
///
/// @param n Number of units to allocate
/// @param unit Size of each unit
/// @param p Address of pointer to hold the result.
/// @sa CeedFree()
int CeedCallocArray(size_t n, size_t unit, void *p) {
  *(void **)p = calloc(n, unit);
  if (n && unit && !*(void **)p)
    return CeedError(NULL, 1, "calloc failed to allocate %zd members of size %zd\n",
                     n, unit);
  return 0;
}

/// Reallocate an array on the host; use CeedRealloc()
///
/// Memory usage can be tracked by the library.
///
/// @param n Number of units to allocate
/// @param unit Size of each unit
/// @param p Address of pointer to hold the result.
/// @sa CeedFree()
int CeedReallocArray(size_t n, size_t unit, void *p) {
  *(void **)p = realloc(*(void **)p, n*unit);
  if (n && unit && !*(void **)p)
    return CeedError(NULL, 1,
                     "realloc failed to allocate %zd members of size %zd\n",
                     n, unit);
  return 0;
}

/// Free memory allocated using CeedMalloc() or CeedCalloc()
///
/// @param p address of pointer to memory.  This argument is of type void* to
/// avoid needing a cast, but is the address of the pointer (which is zeroed)
/// rather than the pointer.
int CeedFree(void *p) {
  free(*(void **)p);
  *(void **)p = NULL;
  return 0;
}

/**
  Wait for a CeedRequest to complete.

  Calling CeedRequestWait on a NULL request is a no-op.

  @param req Address of CeedRequest to wait for; zeroed on completion.
  @return 0 on success
 */
int CeedRequestWait(CeedRequest *req) {
  if (!*req) return 0;
  return CeedError(NULL, 2, "CeedRequestWait not implemented");
}

/// Initialize a \ref Ceed to use the specified resource.
///
/// @param resource  Resource to use, e.g., "/cpu/self"
/// @param ceed The library context
/// @sa CeedRegister() CeedDestroy()
int CeedInit(const char *resource, Ceed *ceed) {
  int ierr;
  size_t matchlen = 0, matchidx;
  unsigned int matchpriority = 100, priority;

  if (!resource) return CeedError(NULL, 1, "No resource provided");
  for (size_t i=0; i<num_backends; i++) {
    size_t n;
    const char *prefix = backends[i].prefix;
    for (n = 0; prefix[n] && prefix[n] == resource[n]; n++) {}
    priority = backends[i].priority;
    if (n > matchlen || (n == matchlen && matchpriority > priority)) {
      matchlen = n;
      matchpriority = priority;
      matchidx = i;
    }
  }
  if (!matchlen) return CeedError(NULL, 1, "No suitable backend");
  ierr = CeedCalloc(1,ceed); CeedChk(ierr);
  (*ceed)->Error = CeedErrorAbort;
  (*ceed)->refcount = 1;
  (*ceed)->data = NULL;
  ierr = backends[matchidx].init(resource, *ceed); CeedChk(ierr);
  return 0;
}

/**
  Destroy a Ceed context

  @param ceed Address of Ceed context to destroy
  @return 0 on success
 */
int CeedDestroy(Ceed *ceed) {
  int ierr;

  if (!*ceed || --(*ceed)->refcount > 0) return 0;
  if ((*ceed)->Destroy) {
    ierr = (*ceed)->Destroy(*ceed); CeedChk(ierr);
  }
  ierr = CeedFree(ceed); CeedChk(ierr);
  return 0;
}
