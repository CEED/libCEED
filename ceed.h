// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
// reserved. See files LICENSE and NOTICE for details.
//
// This file is part of CEED, a collection of benchmarks, miniapps, software
// libraries and APIs for efficient high-order finite element and spectral
// element discretizations for exascale applications. For more information and
// source code availability see http://github.com/ceed.
//
// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
// a collaborative effort of two U.S. Department of Energy organizations (Office
// of Science and the National Nuclear Security Administration) responsible for
// the planning and preparation of a capable exascale ecosystem, including
// software, applications, hardware, advanced system engineering and early
// testbed platforms, in support of the nation's exascale computing imperative.

#ifndef _ceed_h
#define _ceed_h

#ifdef __cplusplus
#  define CEED_EXTERN extern "C"
#else
#  define CEED_EXTERN extern
#endif

#include <stdint.h>
#include <stddef.h>
#include <stdarg.h>
#include <stdio.h>

// We can discuss ways to avoid forcing these to be compile-time decisions, but let's leave that for later.
typedef int32_t CeedInt;
typedef double CeedScalar;

typedef struct Ceed_private *Ceed;
typedef struct CeedRequest_private *CeedRequest;
typedef struct CeedVector_private *CeedVector;
typedef struct CeedElemRestriction_private *CeedElemRestriction;
typedef struct CeedBasis_private *CeedBasis;
typedef struct CeedQFunction_private *CeedQFunction;
typedef struct CeedOperator_private *CeedOperator;

CEED_EXTERN int CeedRegister(const char *prefix, int (*init)(const char *,
                             Ceed));

CEED_EXTERN int CeedInit(const char *resource, Ceed *ceed);
CEED_EXTERN int CeedErrorReturn(Ceed, const char *, int, const char *, int,
                                const char *, va_list);
CEED_EXTERN int CeedErrorAbort(Ceed, const char *, int, const char *, int,
                               const char *, va_list);
CEED_EXTERN int CeedSetErrorHandler(Ceed,
                                    int (*)(Ceed, int, const char *, va_list));
CEED_EXTERN int CeedErrorImpl(Ceed, const char *, int, const char *, int,
                              const char *, ...);
#define CeedError(ceed, ecode, ...)                                     \
  CeedErrorImpl((ceed), __FILE__, __LINE__, __func__, (ecode), __VA_ARGS__)
CEED_EXTERN int CeedDestroy(Ceed *ceed);
CEED_EXTERN int CeedCompose(int n, const Ceed *ceeds, Ceed *composed);

typedef enum {CEED_MEM_HOST, CEED_MEM_DEVICE} CeedMemType;
/* When ownership of dynamically alocated CEED_MEM_HOST pointers is transferred
   to the library (CEED_OWN_POINTER mode), they will be deallocated by calling
   the standard C library function, free(). In particular, pointers allocated
   with the C++ operator new should not be used with CEED_OWN_POINTER mode. */
typedef enum {CEED_COPY_VALUES, CEED_USE_POINTER, CEED_OWN_POINTER} CeedCopyMode;

/* The CeedVectorGet* and CeedVectorRestore* functions provide access to array
   pointers in the desired memory space. Pairing get/restore allows the Vector
   to track access, thus knowing if norms or other operations may need to be
   recomputed. */
CEED_EXTERN int CeedVectorCreate(Ceed ceed, CeedInt len, CeedVector *vec);
CEED_EXTERN int CeedVectorSetArray(CeedVector vec, CeedMemType mtype,
                                   CeedCopyMode cmode, CeedScalar *array);
CEED_EXTERN int CeedVectorGetArray(CeedVector vec, CeedMemType mtype,
                                   CeedScalar **array);
CEED_EXTERN int CeedVectorGetArrayRead(CeedVector vec, CeedMemType mtype,
                                       const CeedScalar **array);
CEED_EXTERN int CeedVectorRestoreArray(CeedVector vec, CeedScalar **array);
CEED_EXTERN int CeedVectorRestoreArrayRead(CeedVector vec,
    const CeedScalar **array);
CEED_EXTERN int CeedVectorDestroy(CeedVector *vec);

/* When CEED_REQUEST_IMMEDIATE is passed as the CeedRequest pointer to a call,
   the called function must ensure that all output is immediately available
   after it returns. In other words, the operation does not need to be executed
   asynchronously, and if it is, the called function will wait for the
   asynchronous execution to complete before returning. */
CEED_EXTERN CeedRequest *CEED_REQUEST_IMMEDIATE;
/* When CEED_REQUEST_NULL (or simply NULL) is given as the CeedRequest pointer
   to a function call, the caller is indicating that he/she will not need to
   call CeedRequestWait to wait for the completion of the operation. In general,
   the operation is expected to be executed asyncronously and its result to be
   available before the execution of next asynchronous operation using the same
   Ceed. */
#define CEED_REQUEST_NULL ((CeedRequest *)NULL)
CEED_EXTERN int CeedRequestWait(CeedRequest *req);

typedef enum {CEED_NOTRANSPOSE, CEED_TRANSPOSE} CeedTransposeMode;

/**
  @brief Create a CeedElemRestriction

  @param ceed       A Ceed object where the CeedElemRestriction will be created.
  @param nelements  Number of elements described in the @a indices array.
  @param esize      Size (number of unknowns) per element.
  @param ndof       The total size of the input CeedVector to which the
                    restriction will be applied. This size may include data
                    used by other CeedElemRestriction objects describing
                    different types of elements.
  @param mtype      Memory type of the @a indices array, see CeedMemType.
  @param cmode      Copy mode for the @a indices array, see CeedCopyMode.
  @param indices    A 2D array of dimensions (@a esize x @a nelements) using
                    column-major storage layout. Column i holds the ordered list
                    of the indices (into the input CeedVector) for the unknowns
                    corresponding to element i, where 0 <= i < @a nelements.
                    All indices must be in the range [0, @a ndof).
  @param r          The address of the variable where the newly created
                    CeedElemRestriction will be stored.

  @return An error code: 0 - success, otherwise - failure.
 */
CEED_EXTERN int CeedElemRestrictionCreate(Ceed ceed, CeedInt nelements,
    CeedInt esize, CeedInt ndof, CeedMemType mtype, CeedCopyMode cmode,
    const CeedInt *indices, CeedElemRestriction *r);

/**
  @brief Create a blocked CeedElemRestriction

  @param ceed        A Ceed object where the CeedElemRestriction will be created.
  @param nelements   Number of elements described ...
  @param esize       Size (number of unknowns) per element.
  @param blocksize   ...
  @param mtype       Memory type of the @a blkindices array, see CeedMemType.
  @param cmode       Copy mode for the @a blkindices array, see CeedCopyMode.
  @param blkindices  ...
  @param r           The address of the variable where the newly created
                     CeedElemRestriction will be stored.

  @return An error code: 0 - success, otherwise - failure.
 */
CEED_EXTERN int CeedElemRestrictionCreateBlocked(Ceed ceed, CeedInt nelements,
    CeedInt esize, CeedInt blocksize, CeedMemType mtype, CeedCopyMode cmode,
    CeedInt *blkindices, CeedElemRestriction *r);
CEED_EXTERN int CeedElemRestrictionApply(CeedElemRestriction r,
    CeedTransposeMode tmode, CeedInt ncomp, CeedTransposeMode lmode, CeedVector u,
    CeedVector ru, CeedRequest *request);
CEED_EXTERN int CeedElemRestrictionDestroy(CeedElemRestriction *r);

// The formalism here is that we have the structure
//   \int_\Omega v^T f_0(u, \nabla u, qdata) + (\nabla v)^T f_1(u, \nabla u, qdata)
// where gradients are with respect to the reference element.

typedef enum {CEED_EVAL_NONE   = 0,
              CEED_EVAL_INTERP = 1, // values at quadrature points
              CEED_EVAL_GRAD   = 2, // gradients
              CEED_EVAL_DIV    = 4, // divergence
              CEED_EVAL_CURL   = 8, // curl
              CEED_EVAL_WEIGHT = 16, // quadrature weights for reference element
             } CeedEvalMode;
typedef enum {CEED_GAUSS = 0, CEED_GAUSS_LOBATTO = 1} CeedQuadMode;

CEED_EXTERN int CeedBasisCreateTensorH1Lagrange(Ceed ceed, CeedInt dim,
    CeedInt ndof, CeedInt P, CeedInt Q, CeedQuadMode qmode, CeedBasis *basis);
CEED_EXTERN int CeedBasisCreateTensorH1(Ceed ceed, CeedInt dim, CeedInt ndof,
                                        CeedInt P1d, CeedInt Q1d, const CeedScalar *interp1d, const CeedScalar *grad1d,
                                        const CeedScalar *qref1d, const CeedScalar *qweight1d, CeedBasis *basis);
CEED_EXTERN int CeedBasisView(CeedBasis basis, FILE *stream);
CEED_EXTERN int CeedBasisApply(CeedBasis basis, CeedTransposeMode tmode,
                               CeedEvalMode emode, const CeedScalar *u, CeedScalar *v);
CEED_EXTERN int CeedBasisGetNumNodes(CeedBasis basis, CeedInt *P);
CEED_EXTERN int CeedBasisGetNumQuadraturePoints(CeedBasis basis, CeedInt *Q);
CEED_EXTERN int CeedBasisDestroy(CeedBasis *basis);

CEED_EXTERN int CeedGaussQuadrature(CeedInt Q, CeedScalar *qref1d,
                                    CeedScalar *qweight1d);
CEED_EXTERN int CeedLobattoQuadrature(CeedInt Q, CeedScalar *qref1d,
                                      CeedScalar *qweight1d);

CEED_EXTERN int CeedQFunctionCreateInterior(Ceed ceed, CeedInt vlength,
    CeedInt nfields, size_t qdatasize, CeedEvalMode inmode, CeedEvalMode outmode,
    int (*f)(void *ctx, void *qdata, CeedInt nq, const CeedScalar *const *u,
             CeedScalar *const *v), const char *focca, CeedQFunction *qf);
CEED_EXTERN int CeedQFunctionSetContext(CeedQFunction qf, void *ctx,
                                        size_t ctxsize);
CEED_EXTERN int CeedQFunctionApply(CeedQFunction qf, void *qdata, CeedInt Q,
                                   const CeedScalar *const *u,
                                   CeedScalar *const *v);
CEED_EXTERN int CeedQFunctionDestroy(CeedQFunction *qf);

CEED_EXTERN int CeedOperatorCreate(Ceed ceed, CeedElemRestriction r,
                                   CeedBasis b, CeedQFunction qf, CeedQFunction dqf, CeedQFunction dqfT,
                                   CeedOperator *op);
CEED_EXTERN int CeedOperatorGetQData(CeedOperator op, CeedVector *qdata);
CEED_EXTERN int CeedOperatorApply(CeedOperator op, CeedVector qdata,
                                  CeedVector ustate, CeedVector residual, CeedRequest *request);
CEED_EXTERN int CeedOperatorApplyJacobian(CeedOperator op, CeedVector qdata,
    CeedVector ustate, CeedVector dustate, CeedVector dresidual,
    CeedRequest *request);
CEED_EXTERN int CeedOperatorDestroy(CeedOperator *op);

static inline CeedInt CeedPowInt(CeedInt base, CeedInt power) {
  CeedInt result = 1;
  while (power) {
    if (power & 1) result *= base;
    power >>= 1;
    base *= base;
  }
  return result;
}

#endif
