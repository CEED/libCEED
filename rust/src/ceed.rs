// #![allow(non_upper_case_globals)]
// #![allow(non_camel_case_types)]
// #![allow(non_snake_case)]
// // #[macro_use(s)]
// // extern crate ndarray;
// use std::fmt;
// use std::ffi::CString;
// use std::io::prelude::*;
// use std::io::{BufReader, SeekFrom};

// use cfile;
// // use std::io::{self, Write};

// mod rust_ceed {
//   include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
// }
// mod ceed_vector;

// /// Ceed context wrapper
// pub struct Ceed {
//   backend : String,
//   // Pointer to C object
//   ceed_ptr : rust_ceed::Ceed,
// }

// /// Display
// impl fmt::Display for Ceed {
//   fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//     write!(f, "{}", self.backend)

//     //let tmpout = cfile::tmpfile().unwrap();
//     //rust_ceed::CeedView(self.ceed_ptr, libc::stdout);
//     //tmpout.seek(SeekFrom::Start(0)).unwrap();
//     //let mut r = BufReader::new(tmpout);
//     //let mut s = String::new();
//     // r.read(&mut s).unwrap();
//     //write!(f, "{}", s)
//   }
// }

// /// Destructor
// impl Drop for Ceed {
//   fn drop(&mut self) {
//     unsafe {
//       rust_ceed::CeedDestroy(&mut self.ceed_ptr);
//     }
//   }
// }

// enum MemType {
//   Host,
//   Device,
// }

// enum CopyMode {

// }

// // Object constructors
// impl Ceed {
//   /// Returns a Ceed context initalized with the specified resource
//   ///
//   /// # arguments
//   ///
//   /// * 'resource' - Resource to use, e.g., "/cpu/self"
//   /// 
//   /// 
//   /// let ceed = init_ceed("/cpu/self/ref/serial");
//   /// 
//   pub fn init(resource: &str) -> Self {
//     // Convert to C string
//     let c_resource = CString::new(resource).expect("CString::new failed");
    
//     // Call to libCEED
//     unsafe {
//       let mut ceed_ptr: rust_ceed::Ceed = libc::malloc(mem::size_of::<rust_ceed::Ceed>()) as rust_ceed::Ceed;
//       rust_ceed::CeedInit(c_resource.as_ptr() as *const i8, &mut ceed_ptr);
//       Ceed { backend : resource.to_string(), ceed_ptr: ceed_ptr }
//     }
//   }

//   /// Vector
//   pub fn vector(&self, n: i32) -> ceed_vector {
//     let mut vec_ptr = libc::malloc(mem::size_of::<rust_ceed::CeedVector>()) as rust_ceed::CeedVector;
//     rust_ceed::CeedVectorCreate(self.ceed_ptr, n, &mut vec_ptr);
//     ceed_vector::CeedVector { ceed_reference: self, ceed_vec_ptr: vec_ptr }
//   }

//   /// Elem Restriction
//   pub fn elem_restriction(&self ceed, nelem : i32, elemsize : i32, ncomp : i32, 
//     compstride : i32, lsize : i32, mtype : MemType, cmode : CopyMode,
//     offsets : vec<i32> &) -> ceed_elem_restriction {

//   }

//   /// Basis

//   /// QFunction

//   /// Operator
// }

// /*
// use std::io::prelude::*;
// use std::io::{BufReader, SeekFrom};
// use cfile::CFile;

// let mut f = CFile::open_tmpfile().unwrap(); // open a tempfile

// assert_eq!(f.write(b"test").unwrap(), 4); // write something to the stream

// f.flush().unwrap(); // force to flush the stream

// assert_eq!(f.seek(SeekFrom::Start(0)).unwrap(), 0); // seek to the beginning of stream

// let mut r = BufReader::new(f);
// let mut s = String::new();
// assert_eq!(r.read_line(&mut s).unwrap(), 4); // read back the text
// assert_eq!(s, "test");
// */



// // use std::fs::File;
// // use std::io::BufReader;
// // use std::io::prelude::*;

// /*
// fn main() -> std::io::Result<()> {
//   let file = File::open("foo.txt")?;
//   let mut buf_reader = BufReader::new(file);
//   let mut contents = String::new();
//   buf_reader.read_to_string(&mut contents)?;
//   assert_eq!(contents, "Hello, world!");
//   Ok(())
// }
// */
// //CeedView(Ceed ceed, FILE *stream)

