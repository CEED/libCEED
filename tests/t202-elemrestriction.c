/// @file
/// Test creation, use, and destruction of a blocked element restriction
/// \test Test creation, use, and destruction of a blocked element restriction
#include <ceed.h>

int main(int argc, char **argv) {
  Ceed ceed;
  CeedVector x, y;
  CeedInt ne = 8;
  CeedInt blksize = 5;
  CeedInt ind[2*ne];
  CeedScalar a[ne+1];
  CeedElemRestriction r;

  CeedInit(argv[1], &ceed);

  CeedVectorCreate(ceed, ne+1, &x);
  for (CeedInt i=0; i<ne+1; i++)
    a[i] = 10 + i;
  CeedVectorSetArray(x, CEED_MEM_HOST, CEED_USE_POINTER, a);

  for (CeedInt i=0; i<ne; i++) {
    ind[2*i+0] = i;
    ind[2*i+1] = i+1;
  }
  CeedElemRestrictionCreateBlocked(ceed, ne, 2, blksize, ne+1, 1, CEED_MEM_HOST,
                                   CEED_USE_POINTER, ind, &r);
  CeedVectorCreate(ceed, 2*blksize*2, &y);
  CeedVectorSetValue(y, 0); // Allocates array

  // NoTranspose
  CeedElemRestrictionApply(r, CEED_NOTRANSPOSE, CEED_NOTRANSPOSE, x, y,
                           CEED_REQUEST_IMMEDIATE);
  CeedVectorView(y, "%12.8f", stdout);

  // Transpose
  CeedVectorGetArray(x, CEED_MEM_HOST, (CeedScalar **)&a);
  for (CeedInt i=0; i<ne+1; i++)
    a[i] = 0;
  CeedVectorRestoreArray(x, (CeedScalar **)&a);
  CeedElemRestrictionApply(r, CEED_TRANSPOSE, CEED_NOTRANSPOSE, y, x,
                           CEED_REQUEST_IMMEDIATE);
  CeedVectorView(x, "%12.8f", stdout);

  CeedVectorDestroy(&x);
  CeedVectorDestroy(&y);
  CeedElemRestrictionDestroy(&r);
  CeedDestroy(&ceed);
  return 0;
}
