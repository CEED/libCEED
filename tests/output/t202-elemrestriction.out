CeedVector length 20
   10.00000000
   11.00000000
   12.00000000
   13.00000000
   14.00000000
   11.00000000
   12.00000000
   13.00000000
   14.00000000
   15.00000000
   15.00000000
   16.00000000
   17.00000000
   17.00000000
   17.00000000
   16.00000000
   17.00000000
   18.00000000
   18.00000000
   18.00000000
CeedVector length 9
   10.00000000
   22.00000000
   24.00000000
   26.00000000
   28.00000000
   30.00000000
   32.00000000
   34.00000000
   18.00000000
